CON-84, CON-151: Add to Avail System Overview

DannyS03 opened the pull request:
Elaborating on Avail's design and approach.

Will follow other articles on consensus, Kate commitments (and more).
Will also be good to provide more technicalities for each section in their own doc, supported by diagrams

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+Currently, monolithic blockchain architectures like that of Ethereum 
+cannot efficiently handle execution, settlement, and data availability. 
+
+Using modularity to scale blockchains is what the rollup-centric or 
+application-specific chain model attempt to do. Sometimes, this works 
+best when the settlement and data availability layer are on the same 
+layer, which is the approach Ethereum rollups take.
```
* Rollups modularize execution (execution is off-chain), but leave data availability in the monolithic chain (they submit transaction data to the L1). So they modularize _in part_.
* I don't think rollups "work best" necessarily - they make some tradeoffs, there are pros/cons. The rollup construction can be more secure (depending on the security of the data availability layer), but it is inherently more difficult to scale.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+However, a granular design is designing different layers to be lightweight 
+protocols, like microservices. Then, the overall network becomes a collection 
+of loosely-coupled lightweight protocols. An example is a data availability 
+layer that only specializes in data availability. Polygon Avail is a Substrate-based 
+application-specific chain for DA (data availability). 
```
Hmm I can see what you're trying to say, but I think it's confusing to describe Avail as an app chain, because app chains can be built _on top_ of Avail. So I think it would be better not to refer to Avail in that way.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+
+:::note Substrate runtime
+
+Modifications have been made to the actual Substrate runtime as opposed to 
+the standard plug-and-play approach Substrate's FRAME offers. Please see the 
+[Avail architecture guide] for more details.
```
I'm not familiar with FRAME, but I would phrase this as: Although Avail is based on the Substrate codebase, it includes modifications to the block structure that prevent it from interoperating with other Substrate networks. Avail implements an independent network unrelated to Polkadot or Kusama.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+
+:::
+
+Avail provides a high guarantee of data availability to any light client. 
```
To be more precise:
* Avail nodes do not make higher guarantees to light clients about DA than any other network.
* What Avail does provide is a way for light clients to "sample" small pieces of the block to verify it is actually there.
* Thus, the DA guarantee is something the client determines for itself—it doesn't have to trust nodes.

Also:
* As the number of light clients grows, they collectively sample the entire block (even though each client only samples a small %)
* Light clients actually form a P2P network amongst themselves
* Thus, after a block has been sampled, it becomes _highly available_—that is, even if the nodes were to go down (or attempt to censor a block), the light clients would be able to reconstruct the block by sharing the pieces amongst themselves.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
 sample small amounts of the block data to verify its full
-availability.
+availability. However, there are fundamentally different primitives than 
+standard fraud-proof-based DA systems, which are explained [here].
```
Remove "standard" — fraud proofs are not really more standard than what we do.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+approaches, but the overall goal is to maintain high security while being able 
+to scale. Avail will take rollups to the next level as chains can allocate 
+their data availability component to Avail. Avail also provides an alternative 
+way to bootstrap [any] standalone chain, as chains can offload their data 
+availability. Transaction costs are also reduced; the primary cost 
```
...as sovereign chains can be created that rely on Avail's validators to reach consensus on transaction data and ordering.

vthunder replied:
As for the tx costs point - it's true, but the way it is phrased is very much about validiums, whereas the point about bootstrapping sovereign rollups is about Avail-native chains. These are very different ways to use Avail, so I'd like to put more distance between them / make it clear these are different use cases.

Fundamentally, the reason Avail can remain cheap is that it is able to grow block size with a smaller impact to validator workload than a monolithic chain. When a monolithic chain increases block size, validators have to do a lot more work, because blocks have to be executed and state has to be calculated. Since avail has no execution environment at all, it's much cheaper to increase block size. (Not zero because we do need to calculate KZG commitments and generate proofs — but cheaper).

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+rollups a possibility. Sovereign rollups on Avail allow for seamless upgrades, 
+as updates can be pushed to application-specific nodes to upgrade the chain and, 
+in turn, upgrade to new settlement logic. Whereas in a traditional environment, 
+direct updates would need to be made to contracts, or the network requires a fork.
```
I'd remove 'contracts' here - we're talking about fundamental network upgrades, not contracts.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+what that data is. The data can be transactions but can take
+on other forms too. Data availability on Avail is available 
+for a window of time that it is required. For instance, beyond needing
+data or reconstruction, security is not compromised.
```
*for

vthunder replied:
This point about retention and the next paragraph about storage are good points, but shouldn't be mixed with the point about Avail being agnostic as to what the data is. Should be a separate point imo.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
-  are encouraged to use Avail light clients. However, users are able to run Avail full nodes
-  and it is well supported.
+Typically, Ethereum-like ecosystems contain three types of peers:
 
-:::
+* **Validator nodes:**
+  A validator collects transactions from the mempool, executes them, and generates a 
+  candidate block that is appended to the network. The block contains a small block 
+  header with the digest and metadata of the transactions in the block.
+* **Full nodes:**
+  The candidate block propogates to full nodes across the network for verification. 
+  The nodes will re-execute the transactions contained in the candidate block. 
+* **Light clients**
+  Light clients only fetches the block header to use for verification, and will fetch 
+  transaction details from neighboring full nodes as needed.
```
Why not just talk about what the Avail nodes/clients are? Why talk about Ethereum?

For example - the Avail validators do not execute transactions. They simply package up the transactions (whatever they are) and construct candidate blocks, generating KZG commitments for the data. Other validators check that generated blocks are correct. Similarly, Avail full nodes do not execute anything.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,68 +14,135 @@ image: https://matic.network/banners/matic-network-16x9.png
+
+:::info The goal of Avail is to not be reliant on full nodes to keep data available
+
+  The aim is to give the same DA guarantees to a light client as a full node. 
```
similar, not same

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -84,5 +151,39 @@ use-cases. One such bridge being designed at this time is an
+
+### Block verification -> DA verification
+
+Instead of Avail validators verifying the application state, they concentrate 
+on ensuring the availability of posted transaction data and provides transaction 
+ordering. A block is considered valid only if the data behind that block is available.
```
Not really wrong, but here's some additional context:

* Avail validators take incoming transactions, order them, construct a block, and propose it.
* The block contains special features especially for DA—erasure coding, and KZG commitments. And it's in a special format so that clients are able to do random sampling later, as well as being able to download only a single application's transactions.
* Anyway - they propose a block, and other validators verify the block is correct - well formed, KZG commitments check out, data is there, etc.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -84,5 +151,39 @@ use-cases. One such bridge being designed at this time is an
+
+Requiring data to be available prevents block producers from releasing block headers
+without releasing the data behind them, as this prevents clients from reading the 
+transactions necessary to compute the state of their applications. Avail uses data 
+availability verification to address this through DA checks which utilize erasure 
+codes; these checks are heavily used in data redundancy design.
```
* The DA verification is performed by clients via sampling. It's not something validators or full nodes do, exactly.
* Erasure codes effectively duplicate data in such a way that if part of a block is suppressed, clients can reconstruct that part by using another part of the block. This means that a node trying to suppress that part would actually need to suppress _a lot more_. The technique is used for example in CD-ROMs, and in multi-disk (RAID) arrays (so for example if a hard drive dies, it can be replaced and re-constructed from the data on other disks).

vthunder replied:
I mean—if data was indeed unavailable, then validators would not sign off on a block, that is true—but it's also true of any other chain. The thing that makes Avail unique isn't that _validators_ check the data, but rather that the chain is structured so that _anyone_ can check DA without actually downloading the data.

vthunder commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -84,5 +151,39 @@ use-cases. One such bridge being designed at this time is an
+
+### Transaction settlement
+
+Avail will use a settlement layer built with Polygon Edge. The settlement layer 
+provides an EVM-compatible blockchain for rollups to store their data and perform 
+dispute resolution. The settlement layer utilizes Polygon Avail for its own DA. 
+When rollups are using a settlement layer, they also inherit all the 
+DA properties of Avail.
```
This is all correct - but note that there are different ways to use Avail, and the validium use case (referenced way above) will not use this settlement layer (validiums will settle on Ethereum). Just FYI!

vthunder reviewed (dismissed):
Left a bunch of comments - please review them and make tweaks as necessary (some point out issues, others are more FYI to add context for you, in case you want to frame things differently or explain further).

DannyS03 commented:
@vthunder many thanks for the review. Addressed the feedback and areas that were not directly addressed were a result of the suggestions already existing in another part of the doc

DannyS03 commented:
@Nazeim @cerberushades can review

cerberushades commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
-  - Download headers, but not full blocks
-  - Randomly sample small parts of the block to verify availability
-  - Expose a local API to interact with the Avail network
+   Clients that only download block headers randomly sample small parts of 
```
```suggestion
   Clients that only download block headers and random samples of 
```

DannyS03 replied:
suggestion doesn't work with how random sampling is supposed to be mentioned

cerberushades commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
+:::info The goal of Avail is not to be reliant on full nodes to keep data available
+
+  The aim is to give similar DA guarantees to a light client as a full node. 
+  Users are encouraged to use Avail light clients. However, users can run Avail 
```
```suggestion
  Users are encouraged to use Avail light clients. However, they can still run Avail 
```

cerberushades commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
-A more detailed overview of the Avail system can be found in the
-this [Introduction to Avail by Polygon blog post](https://medium.com/the-polygon-blog/introducing-avail-by-polygon-a-robust-general-purpose-scalable-data-availability-layer-98bc9814c048).
+- [Introduction to Avail by Polygon blog post](https://medium.com/the-polygon-blog/introducing-avail-by-polygon-a-robust-general-purpose-scalable-data-availability-layer-98bc9814c048).
+- [Polygon Talks: Polygon Avail]
```
no link

DannyS03 replied:
updated

DannyS03 commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
-A more detailed overview of the Avail system can be found in the
-this [Introduction to Avail by Polygon blog post](https://medium.com/the-polygon-blog/introducing-avail-by-polygon-a-robust-general-purpose-scalable-data-availability-layer-98bc9814c048).
+- [Introduction to Avail by Polygon blog post](https://medium.com/the-polygon-blog/introducing-avail-by-polygon-a-robust-general-purpose-scalable-data-availability-layer-98bc9814c048).
+- [Polygon Talks: Polygon Avail]
```
```suggestion
- [Polygon Talks: Polygon Avail](https://www.youtube.com/watch?v=okqMT1v3xi0)
```

DannyS03 commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
 sample small amounts of the block data to verify its full
-availability.
+availability. However, there are fundamentally different primitives than 
+fraud-proof-based DA systems, which are explained [here].
```
```suggestion
fraud-proof-based DA systems, which are explained [here](https://blog.polygon.technology/the-data-availability-problem-6b74b619ffcc/).
```

DannyS03 commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
-downloading the whole block by leveraging Kate polynomial commitments,
-erasure coding, and other technologies to allow light clients (which
+Currently, monolithic blockchain architectures like that of Ethereum 
+cannot efficiently handle the execution, settlement, and data availability. 
```
can leave "the" out

DannyS03 commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
 sample small amounts of the block data to verify its full
-availability.
+availability. However, there are fundamentally different primitives than 
+fraud-proof-based DA systems, which are explained [here].
```
updated

DannyS03 commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
+
+Avail will take rollups to the next level as chains can allocate 
+their data availability component to Avail. Avail also provides an alternative 
+way to bootstrap [any] standalone chain, as chains can offload their data 
```
```suggestion
way to bootstrap any standalone chain, as chains can offload their data 
```

DannyS03 commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
+
+Avail will take rollups to the next level as chains can allocate 
+their data availability component to Avail. Avail also provides an alternative 
+way to bootstrap [any] standalone chain, as chains can offload their data 
```
updated

DannyS03 commented on docs/avail/architecture/avail-system-overview.md:
```diff
@@ -14,75 +14,229 @@ image: https://matic.network/banners/matic-network-16x9.png
-what that data is. The data can of course be transactions, but can take
-on other forms too.
+Avail guarantees that block data is available but does not care about
+what that data is. The data can be transactions but can take
```
we are being specific about transaction data

DannyS03 commented:
@cerberushades addressed feedback

DannyS03 commented:
@cerberushades checking if we'd like to change anything else
